macro/src/class_info.rs

```diff
@@ -1,73 +1,168 @@
+    pub classes: BTreeMap<DotId, Arc<ClassInfo>>,
+}
+
+impl RootMap {
+    /// Finds the class with the given name (if present).
+    pub fn find_class(&self, cn: &DotId) -> Option<&Arc<ClassInfo>> {
+        self.classes.get(cn)
+    }
+
+    /// Finds the package with the given name (if present).
+    pub fn find_package(&self, ids: &[Id]) -> Option<&SpannedPackageInfo> {
+        let (p0, ps) = ids.split_first().unwrap();
+        self.subpackages.get(p0)?.find_subpackage(ps)
+    }
+
+    pub fn to_packages(&self) -> impl Iterator<Item = &SpannedPackageInfo> {
+        self.subpackages.values()
+    }
+
+    /// Find the names of all classes contained within.
+    pub fn class_names(&self) -> Vec<DotId> {
+        self.classes.keys().cloned().collect()
+    }
+}
+
+#[derive(Debug)]
 pub struct SpannedPackageInfo {
     pub name: Id,
     pub span: Span,
     pub subpackages: BTreeMap<Id, SpannedPackageInfo>,
-    pub classes: Vec<SpannedClassInfo>,
+    pub classes: Vec<DotId>,
 }
 
-pub struct SpannedClassInfo {
-    /// The complete class info loaded from javap
-    pub info: ClassInfo,
+impl SpannedPackageInfo {
+    /// Find a (sub)package given its relative name
+    pub fn find_subpackage(&self, ids: &[Id]) -> Option<&SpannedPackageInfo> {
+        let Some((p0, ps)) = ids.split_first() else {
+            return Some(self);
+        };
 
-    /// The span where user declared interest in this class
-    pub span: Span,
+        self.subpackages.get(p0)?.find_subpackage(ps)
+    }
 
-    /// The listing of members user wants to include
-    pub members: MemberListing,
+    /// Finds a class in this package with the given name (if any)
+    pub fn find_class(&self, cn: &Id) -> Option<&DotId> {
+        self.classes.iter().find(|c| c.is_class(cn))
+    }
 }
 
-impl SpannedClassInfo {
-    pub fn parse(t: &str, span: Span, members: MemberListing) -> Result<Self, SpanError> {
-        match javap::parse_class_info(&t) {
-            Ok(info) => Ok(SpannedClassInfo {
-                span,
-                info,
-                members,
-            }),
-            Err(message) => Err(SpanError { span, message }),
-        }
-    }
+#[derive(Debug)]
+pub enum ClassDecl {
+    /// User wrote `class Foo { * }`
+    Reflected(ReflectedClassInfo),
+
+    /// User wrote `class Foo { ... }` with full details.
+    Specified(ClassInfo),
 }
 
-impl Parse for SpannedClassInfo {
+impl Parse for ClassDecl {
     fn parse(p: &mut crate::parse::Parser) -> Result<Option<Self>, SpanError> {
-        let Some(t) = p.eat_string_literal() else {
+        // Look for a keyword that could start a class definition.
+        let Some(t0) = p.peek_token() else {
             return Ok(None);
         };
-        let span = p.last_span().unwrap();
-        let r = Self::parse(&t, span, MemberListing::all())?;
+        match t0 {
+            TokenTree::Ident(i) => {
+                static START_KEYWORDS: &[&str] = &["class", "public", "final", "abstract"];
```

Comment:
Future looking: we'll want to support for `interface`, `enum`, and `record` at some point.

Reply:
Fixed